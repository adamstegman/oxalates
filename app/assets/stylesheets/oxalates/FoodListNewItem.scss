@import "./body-form";
@import "./error";

.add-food {
  @extend %body-form;
}

.error {
  @extend %error;
}
