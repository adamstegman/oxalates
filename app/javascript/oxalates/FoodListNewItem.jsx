import React from 'react';
import PropTypes from 'prop-types';
import { foodPropType } from './foodPropType';

export const FoodListNewItem = ({ newFood, newFoodListId, setNewFood, cancelNewFood, createFood, error, password }) => {
  const setNewFoodName = event => {
    const newFoodWithName = Object.assign({}, newFood, { name: event.target.value });
    setNewFood(newFoodListId, newFoodWithName);
  };
  const setNewFoodOxalateMg = event => {
    const newFoodWithOxalateMg = Object.assign({}, newFood, { oxalateMg: event.target.value });
    setNewFood(newFoodListId, newFoodWithOxalateMg);
  };
  const setNewFoodServing = event => {
    const newFoodWithServing = Object.assign({}, newFood, { serving: event.target.value });
    setNewFood(newFoodListId, newFoodWithServing);
  };
  const _createFood = event => {
    event.preventDefault();
    createFood(password, newFood, newFoodListId);
  };

  const {
    name,
    oxalateMg,
    serving,
  } = newFood;
  let errorMessages = null;
  if (error instanceof Array) {
    errorMessages = error.map((msg, i) => <p key={i} className="error">{msg}</p>);
  } else if (error) {
    errorMessages = <p className="error">{error}</p>;
  }
  return (
    <form className="add-food" onSubmit={_createFood}>
      <label htmlFor="name">New food name</label>
      <input type="text" name="name" autoFocus value={name || ''} onChange={setNewFoodName} />
      <label htmlFor="oxalateMg">Oxalates per serving (mg)</label>
      <input type="number" name="oxalateMg" step="any" value={oxalateMg || 0.0} onChange={setNewFoodOxalateMg} />
      <label htmlFor="serving">Serving</label>
      <input type="text" name="serving" value={serving || ''} onChange={setNewFoodServing} />
      {errorMessages}
      <input type="submit" value="Add" />
      <button type="button" onClick={cancelNewFood}>Cancel</button>
    </form>
  );
};

FoodListNewItem.propTypes = {
  newFood: PropTypes.shape({}).isRequired,
  newFoodListId: PropTypes.node.isRequired,
  cancelNewFood: PropTypes.func.isRequired,
  createFood: PropTypes.func.isRequired,
  setNewFood: PropTypes.func.isRequired,
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  password: PropTypes.string,
};
