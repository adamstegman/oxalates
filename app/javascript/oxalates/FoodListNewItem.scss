@import "./body-form";

.add-food {
  @extend %body-form;
}
