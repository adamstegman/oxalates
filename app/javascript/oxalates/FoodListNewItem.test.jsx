import React from 'react';
import renderer from 'react-test-renderer';

import { FoodListNewItem } from './FoodListNewItem';

test('FoodListNewItem renders an empty form', () => {
  const component = renderer.create(
    <FoodListNewItem newFood={{}} newFoodListId={1} createFood={() => {}} setNewFood={() => {}} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodListNewItem renders a pre-filled form', () => {
  const newFood = {
    name: 'some food',
    serving: 'some serving',
    oxalateMg: 3.14,
  }
  const component = renderer.create(
    <FoodListNewItem newFood={newFood} newFoodListId={1} createFood={() => {}} setNewFood={() => {}} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodListNewItem renders a failed form', () => {
  const newFood = {
    name: 'some food',
    serving: 'some serving',
    oxalateMg: 3.14,
  }
  const component = renderer.create(
    <FoodListNewItem newFood={newFood} newFoodListId={1} createFood={() => {}} setNewFood={() => {}} error={['error 1']} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
