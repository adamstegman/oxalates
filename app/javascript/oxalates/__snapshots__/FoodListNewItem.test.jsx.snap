// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodListNewItem renders a failed form 1`] = `
<form
  className="add-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    New food name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value="some food"
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={3.14}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value="some serving"
  />
  <p
    className="error"
  >
    error 1
  </p>
  <input
    type="submit"
    value="Add"
  />
</form>
`;

exports[`FoodListNewItem renders a pre-filled form 1`] = `
<form
  className="add-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    New food name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value="some food"
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={3.14}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value="some serving"
  />
  <input
    type="submit"
    value="Add"
  />
</form>
`;

exports[`FoodListNewItem renders a server error 1`] = `
<form
  className="add-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    New food name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value="some food"
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={3.14}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value="some serving"
  />
  <p
    className="error"
  >
    error message
  </p>
  <input
    type="submit"
    value="Add"
  />
</form>
`;

exports[`FoodListNewItem renders an empty form 1`] = `
<form
  className="add-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    New food name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={0}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value=""
  />
  <input
    type="submit"
    value="Add"
  />
</form>
`;
